### PR TITLE
MISUV-7725-error-msg-and-save-choice

### PR DESCRIPTION
**Pull request by james-king-hmrc**
error-msg-and-save-choice

### CODE REVIEW

File: app/services/optout/OptOutService.scala [hunk 1]
```diff
@@ -97,41 +92,62 @@ class OptOutService @Inject()(itsaStatusUpdateConnector: ITSAStatusUpdateConnect
...
 
-    val optOutYearsToUpdate = optOutProposition.optOutYearsToUpdate(intent)
+    def makeUpdateCalls(optOutYearsToUpdate: Seq[OptOutTaxYear]): Seq[Future[OptOutUpdateResponse]] = {
+      optOutYearsToUpdate.map(optOutYear => itsaStatusUpdateConnector.requestOptOutForTaxYear(optOutYear.taxYear, user.nino, optOutUpdateReason))
+    }
 
-    val responses: Seq[Future[OptOutUpdateResponse]] = optOutYearsToUpdate.map(optOutYear =>
-      itsaStatusUpdateConnector.requestOptOutForTaxYear(optOutYear.taxYear, user.nino, optOutUpdateReason))
+    def findAnyFailOrFirstSuccess(responses: Seq[OptOutUpdateResponse]): OptOutUpdateResponse = {
+      responses.find(_.isInstanceOf[OptOutUpdateResponseFailure]).getOrElse(responses.head)
```
Comment (avelx) [inline app/services/optout/OptOutService.scala:123]: us of isInstance or asInstance is discourage in general and used as a last option; you should be able to pattern match or use collect etc over the result, depending what you are trying to achieve

File: app/controllers/optOut/ConfirmOptOutController.scala [hunk 2]
```diff
@@ -45,40 +47,41 @@ class ConfirmOptOutController @Inject()(view: ConfirmOptOut,
...
   def show(isAgent: Boolean): Action[AnyContent] = auth.authenticatedAction(isAgent) {
     implicit user =>
       withRecover(isAgent) {
-        withOptOutQualifiedTaxYear(isAgent)(
-          viewModel => Ok(view(viewModel, isAgent = isAgent))
-        )
+
+        val resultToReturn = for {
+          viewModel <- OptionT(optOutService.optOutCheckPointPageViewModel())
+          result <- OptionT(Future.successful(Option(toPropositionView(isAgent, viewModel))))
```
Comment (SiddharthAtos) [inline app/controllers/optOut/ConfirmOptOutController.scala:56]: should fix: use `Some`  eg. `Option(toPropositionView(isAgent, viewModel))` -> `Some(toPropositionView(isAgent, viewModel))` to represent value

File: app/services/optout/OptOutService.scala [hunk 3]
```diff
@@ -97,74 +89,97 @@ class OptOutService @Inject()(itsaStatusUpdateConnector: ITSAStatusUpdateConnect
...
+    }
 
-    val responses: Seq[Future[OptOutUpdateResponse]] = optOutYearsToUpdate.map(optOutYear =>
-      itsaStatusUpdateConnector.requestOptOutForTaxYear(optOutYear.taxYear, user.nino, optOutUpdateReason))
+    val result = for {
+      intentTaxYear <- OptionT(fetchSavedIntent())
+      intentOptOutTaxYear <- OptionT(Future.successful(optOutProposition.optOutTaxYearFor(intentTaxYear)))
+      yearsToUpdate <- OptionT(Future.successful(Option(optOutProposition.optOutYearsToUpdate(intentOptOutTaxYear))))
+      responsesSeqOfFutures  <- OptionT(Future.successful(Option(makeUpdateCalls(yearsToUpdate))))
```
Comment (SiddharthAtos) [inline app/services/optout/OptOutService.scala:127]: should fix: use Some eg. Option(makeUpdateCalls(yearsToUpdate)) -> Some(makeUpdateCalls(yearsToUpdate)) to represent value. 

update similar in following code

File: app/controllers/optOut/ConfirmOptOutController.scala (hunk 2 above)
Comment (james-king-hmrc) [reply app/controllers/optOut/ConfirmOptOutController.scala:56]: won't compile will show you, I know a bit strange

File: app/controllers/optOut/OptOutChooseTaxYearController.scala [hunk 4]
```diff
@@ -45,9 +49,37 @@ class OptOutChooseTaxYearController @Inject()(val optOutChooseTaxYear: OptOutCho
...
 
+  def submit(isAgent: Boolean): Action[AnyContent] = auth.authenticatedAction(isAgent) {
+    implicit user =>
+      optOutService.getTaxYearsAvailableForOptOut().flatMap { availableOptOutTaxYear =>
+        optOutService.getSubmissionCountForTaxYear(availableOptOutTaxYear).map { submissionCountForTaxYear =>
+
+          val onError: Form[ConfirmOptOutMultiTaxYearChoiceForm] => Result = formWithError =>
+            BadRequest(optOutChooseTaxYear(formWithError, availableOptOutTaxYear, submissionCountForTaxYear, isAgent))
+          val onSuccess: ConfirmOptOutMultiTaxYearChoiceForm => Result = form => {
+            saveTaxYearChoice(form)
```
[the remark this answers is not in the document]
Comment (james-king-hmrc) [reply app/controllers/optOut/OptOutChooseTaxYearController.scala:65]: not sure what you mean, let's pair on this

File: app/services/optout/OptOutService.scala (hunk 1 above)
Comment (james-king-hmrc) [reply app/services/optout/OptOutService.scala:120]: discussed this with Siddharth and we can't think of a better way to do this, this is a very specific operation of finding any fail or the first success as the method name suggests

File: app/controllers/optOut/ConfirmOptOutController.scala (hunk 2 above)
Comment (james-king-hmrc) [reply app/controllers/optOut/ConfirmOptOutController.scala:56]: done, have to use Option due to OptionT being Invariance class type

File: app/controllers/optOut/OptOutChooseTaxYearController.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (james-king-hmrc) [reply app/controllers/optOut/OptOutChooseTaxYearController.scala]: done

File: app/services/optout/OptOutService.scala (hunk 3 above)
Comment (james-king-hmrc) [reply app/services/optout/OptOutService.scala:127]: done

File: app/services/optout/OptOutService.scala (hunk 1 above)
Comment (avelx) [reply app/services/optout/OptOutService.scala:120]: you can do something like, also I am not sure if it is guarantied to have head item in the response
```
responses
        .find{
          case _: OptOutUpdateResponseFailure => true
          case _ => false }
        .getOrElse(responses.head)
```

**Review by SiddharthAtos — APPROVED** [review]

**Review by SiddharthAtos — APPROVED** [review]

**Pull request merged**